In [1]:
import pandas as pd													# data manipulation
import matplotlib.pyplot as plt												# graphs
import seaborn as sns													# graphs
import numpy as np													# graphs
from scipy.stats import norm												# statistics
from sklearn.preprocessing import StandardScaler									# statistics
from scipy import stats													# statistics
from math import *													# advance calculation	
from matplotlib import cm												# graphs
import sys														# system functions
import warnings														# indicating warnings
warnings.filterwarnings('ignore')											# ignore by default
from matplotlib.ticker import LinearLocator										# graphs
pd.set_option('display.max_columns', None)										# show all columns
pd.set_option('display.max_rows', None)											# show all rows
np.set_printoptions(threshold=sys.maxsize)										# show arrays full
pd.set_option("display.max_colwidth", -1)										# show a complete text in a cell
pd.set_option('display.precision',5)											# precision decimals pd
np.set_printoptions(precision=3)											# precision decimals np
from xlsxwriter.workbook import Workbook										# CSV writing
import csv														# CSV writing	
import xgboost as xgb													# prediction
import pickle														# prediction	
import sklearn														# mutual information
import statsmodels.api as sm												# coefficient
from sklearn.ensemble import RandomForestClassifier									# random forest
from sklearn import linear_model											# linear regression
import pymysql														# MySql connection
import datetime														# time conversions
import random														# randomchoice
import math as math													# maths
import statistics													# statistics
import glob														# paths and files management
import matplotlib.image as mpimg											# matplot image

In [2]:
df = pd.read_csv('df.csv')
df.drop(columns=['Unnamed: 0'], inplace=True)

In [3]:
# Show null values and duplicates
total = df.isnull().sum().sort_values(ascending=False)
percent = round((df.isnull().sum()/df.isnull().count()).sort_values(ascending=False)*100, 3)
missing_data = pd.concat([total, percent], axis=1, keys=['NullTotal', 'Null%'])
for dup in missing_data.index:
    missing_data.at[dup,'DupTotal'] = df.duplicated(subset=[dup]).sum()
    missing_data.at[dup,'Dup%'] = round((df.duplicated(subset=[dup]).sum())/df.shape[0]*100,3)
missing_data

NullTotal  Null%  DupTotal    Dup%
CoverageRatio%    1          0.119  762.0     90.714
Demand            0          0.000  696.0     82.857
PeopleSaw+1Cars   0          0.000  734.0     87.381
PeopleSaw0Cars    0          0.000  766.0     91.190
RidesPerOnline    0          0.000  686.0     81.667
Rides             0          0.000  783.0     93.214
Busy(H)           0          0.000  804.0     95.714
Waiting(H)        0          0.000  794.0     94.524
Booking(H)        0          0.000  811.0     96.548
OnlinePerDrivers  0          0.000  834.0     99.286
Online(H)         0          0.000  787.0     93.690
Drivers           0          0.000  727.0     86.548
Week              0          0.000  835.0     99.405
Hour              0          0.000  816.0     97.143
Date              0          0.000  805.0     95.833

In [4]:
# Count Values
pd.set_option('display.max_rows', 1000)
for column in df.columns:
    print(column,':')
    display(df[column].value_counts(dropna=False))
    print('___________________________________','\n')	

Date :


16/12/2016    24
12/12/2016    24
17/11/2016    24
15/12/2016    24
04/12/2016    24
18/12/2016    24
25/11/2016    24
03/12/2016    24
19/11/2016    24
15/11/2016    24
29/11/2016    24
10/12/2016    24
13/12/2016    24
24/11/2016    24
16/11/2016    24
02/12/2016    24
14/12/2016    24
21/11/2016    24
20/11/2016    24
30/11/2016    24
05/12/2016    24
14/11/2016    24
17/12/2016    24
11/12/2016    24
18/11/2016    24
01/12/2016    24
22/11/2016    24
27/11/2016    24
07/12/2016    24
06/12/2016    24
28/11/2016    24
26/11/2016    24
23/11/2016    24
08/12/2016    24
09/12/2016    24
Name: Date, dtype: int64

___________________________________ 

Hour :


23    35
22    35
1     35
2     35
3     35
4     35
5     35
6     35
7     35
8     35
9     35
10    35
11    35
12    35
13    35
14    35
15    35
16    35
17    35
18    35
19    35
20    35
21    35
0     35
Name: Hour, dtype: int64

___________________________________ 

Week :


5    168
4    168
3    168
2    168
1    168
Name: Week, dtype: int64

___________________________________ 

Drivers :


72     19
71     19
62     17
65     16
76     16
75     16
74     15
28     14
85     13
40     13
53     13
83     13
56     13
81     13
80     13
24     13
67     13
84     12
34     12
45     12
68     12
70     11
46     11
43     11
29     11
86     11
77     11
61     11
59     11
54     11
51     11
27     11
58     10
23     10
63     10
64     10
69     10
31     9 
25     9 
47     9 
38     8 
42     8 
17     8 
26     8 
90     8 
66     8 
78     8 
91     8 
92     8 
82     7 
87     7 
37     7 
39     7 
41     7 
44     7 
48     7 
20     7 
108    7 
52     7 
95     7 
73     7 
96     7 
104    7 
93     6 
60     6 
32     6 
57     6 
35     6 
36     6 
18     6 
30     5 
79     5 
14     5 
15     5 
102    5 
97     5 
94     5 
106    5 
50     5 
55     5 
98     5 
21     4 
88     4 
22     4 
89     4 
110    4 
16     4 
100    4 
101    4 
49     4 
116    4 
103    4 
113    3 
12     3 
13     2 
19     2 
9      2 
99     2 
109    2 
112    2 


___________________________________ 

Online(H) :


22    37
19    31
27    31
26    29
30    28
23    28
25    28
18    27
10    26
28    25
7     25
21    24
24    24
32    23
11    23
16    23
29    22
14    21
9     20
12    20
6     20
20    20
15    20
34    20
31    19
33    19
35    18
8     17
36    17
13    16
17    14
38    14
40    13
5     12
37    10
4     8 
42    8 
45    7 
39    7 
41    6 
44    5 
43    5 
46    5 
51    4 
50    4 
2     3 
47    3 
48    3 
49    3 
3     2 
53    1 
52    1 
56    1 
Name: Online(H), dtype: int64

___________________________________ 

OnlinePerDrivers :


0.4    463
0.3    274
0.5    75 
0.2    26 
0.6    1  
0.1    1  
Name: OnlinePerDrivers, dtype: int64

___________________________________ 

Booking(H) :


4     91
3     81
1     78
2     78
0     74
5     65
6     58
7     50
8     50
9     32
11    28
12    24
13    20
10    16
14    15
15    14
17    13
16    9 
18    8 
21    8 
22    8 
20    4 
23    4 
24    4 
19    3 
26    2 
27    1 
25    1 
29    1 
Name: Booking(H), dtype: int64

___________________________________ 

Waiting(H) :


9     43
16    41
19    40
14    40
7     39
11    39
13    37
12    35
23    34
5     32
8     31
15    30
6     29
17    28
22    28
18    26
20    25
21    24
24    23
10    22
26    20
25    20
4     20
27    17
28    16
30    15
31    11
29    9 
36    9 
3     7 
2     7 
33    6 
34    5 
41    5 
45    5 
32    4 
42    4 
37    3 
39    2 
43    2 
44    2 
47    1 
35    1 
38    1 
40    1 
48    1 
Name: Waiting(H), dtype: int64

___________________________________ 

Busy(H) :


19    54
11    49
9     47
17    46
15    44
20    44
18    43
12    43
16    43
22    42
14    38
6     37
8     34
10    34
21    32
7     32
23    30
13    27
24    23
5     23
25    12
26    12
4     11
27    10
28    8 
30    5 
3     5 
33    3 
29    2 
2     1 
37    1 
35    1 
31    1 
32    1 
34    1 
1     1 
Name: Busy(H), dtype: int64

___________________________________ 

Rides :


1     52
0     45
6     45
9     40
8     40
4     35
7     34
11    33
10    31
13    30
20    29
3     28
5     27
12    27
16    27
2     24
17    22
15    21
19    20
14    19
24    17
18    16
22    14
21    13
30    12
26    12
27    11
25    10
32    9 
31    9 
29    9 
28    8 
37    8 
23    7 
36    7 
35    6 
43    5 
38    4 
40    4 
34    3 
41    3 
42    3 
47    3 
48    3 
50    2 
33    2 
57    1 
56    1 
53    1 
51    1 
59    1 
49    1 
46    1 
45    1 
44    1 
39    1 
62    1 
Name: Rides, dtype: int64

___________________________________ 

RidesPerOnline :


0.00    45
0.50    23
1.00    23
0.67    17
0.47    16
0.32    16
0.57    14
0.33    14
0.20    13
0.71    13
0.25    12
0.29    12
0.43    12
0.31    12
0.27    12
0.22    11
0.39    11
0.44    10
0.37    10
0.52    10
0.19    10
0.30    10
0.42    10
0.75    10
0.10    9 
0.68    9 
0.59    9 
0.86    9 
0.13    9 
0.35    9 
0.17    9 
0.61    9 
0.28    9 
0.14    9 
0.38    9 
0.24    9 
0.34    8 
0.53    8 
0.65    8 
0.55    8 
0.36    8 
0.69    8 
0.41    7 
0.63    7 
0.79    7 
0.70    7 
0.91    7 
1.11    7 
0.21    7 
0.45    7 
0.64    7 
0.46    7 
0.56    7 
0.88    6 
1.07    6 
0.58    6 
0.40    6 
0.81    6 
0.73    6 
0.78    6 
0.48    6 
0.11    6 
0.62    5 
0.23    5 
0.94    5 
0.95    5 
0.26    5 
0.09    5 
0.89    5 
0.54    5 
0.15    4 
0.60    4 
0.82    4 
0.06    4 
0.08    4 
0.80    4 
1.47    4 
0.16    4 
0.96    4 
1.03    4 
0.93    4 
0.76    4 
1.18    4 
1.25    4 
1.21    3 
1.08    3 
0.07    3 
1.15    3 
1.09    3 
1.27    3 
1.05    3 

___________________________________ 

PeopleSaw0Cars :


2      70
3      70
1      68
5      60
4      59
6      55
7      42
0      40
9      38
8      34
10     32
12     21
11     20
15     15
16     15
13     14
14     13
17     13
18     10
22     10
20     8 
24     7 
21     7 
27     7 
38     5 
43     5 
19     5 
23     5 
34     5 
25     4 
26     4 
28     4 
51     4 
37     4 
32     4 
29     3 
30     3 
39     3 
33     3 
31     3 
36     2 
61     2 
48     2 
69     2 
62     2 
80     2 
59     2 
58     2 
54     2 
66     2 
47     2 
105    2 
44     2 
41     2 
90     1 
67     1 
85     1 
83     1 
96     1 
106    1 
75     1 
87     1 
35     1 
64     1 
63     1 
124    1 
57     1 
52     1 
50     1 
49     1 
46     1 
45     1 
42     1 
129    1 
Name: PeopleSaw0Cars, dtype: int64

___________________________________ 

PeopleSaw+1Cars :


25     26
3      25
34     22
30     22
19     21
4      20
26     20
32     19
33     16
2      16
37     16
49     16
21     15
5      15
44     14
22     14
39     14
29     14
31     13
40     13
46     13
27     13
43     12
28     12
14     12
48     12
41     12
8      11
10     11
52     11
24     11
23     11
36     11
45     10
66     10
50     10
38     10
1      10
18     10
17     10
20     9 
15     9 
13     9 
9      8 
35     8 
6      8 
7      8 
42     8 
47     8 
71     7 
16     7 
12     7 
60     7 
59     7 
55     7 
11     6 
54     6 
63     6 
56     6 
64     6 
53     6 
76     5 
51     5 
57     5 
83     5 
68     5 
62     5 
65     5 
73     4 
72     4 
77     4 
67     4 
61     4 
0      3 
91     3 
97     3 
89     3 
78     3 
79     3 
86     3 
107    3 
58     2 
96     2 
87     2 
90     2 
102    2 
75     2 
74     2 
70     2 
69     2 
81     2 
116    1 
141    1 
127    1 
153    1 
158    1 
115    1 
169    1 
106    1 
105    1 


___________________________________ 

Demand :


35     25
29     18
37     17
26     15
53     15
5      15
44     15
32     15
39     14
38     14
10     14
9      14
30     14
31     14
42     14
6      13
51     13
34     13
28     13
43     12
25     11
62     11
45     11
46     11
40     11
20     11
18     11
8      11
3      11
22     10
49     10
36     10
55     10
65     9 
27     9 
33     9 
74     9 
48     9 
41     9 
15     9 
16     8 
23     8 
24     8 
4      8 
63     8 
7      7 
11     7 
91     7 
21     7 
61     7 
59     7 
58     7 
52     6 
64     6 
66     6 
67     6 
47     6 
70     6 
54     6 
71     6 
12     6 
89     5 
73     5 
60     5 
50     5 
19     5 
14     5 
13     5 
72     5 
83     5 
86     5 
77     5 
78     5 
2      5 
85     4 
87     4 
76     4 
93     4 
80     4 
82     4 
88     4 
56     3 
166    3 
17     3 
123    3 
113    3 
105    3 
57     3 
81     3 
102    3 
68     3 
97     3 
98     3 
100    2 
1      2 
192    2 
69     2 
92     2 
140    2 
101    2 


___________________________________ 

CoverageRatio% :


100.0    39
86.0     35
80.0     28
87.0     27
85.0     26
67.0     26
88.0     26
90.0     23
91.0     23
79.0     23
95.0     22
84.0     21
83.0     21
93.0     21
89.0     20
94.0     20
82.0     18
92.0     17
76.0     16
96.0     15
70.0     15
78.0     15
75.0     14
72.0     14
73.0     13
60.0     13
63.0     13
77.0     13
81.0     13
98.0     12
50.0     12
74.0     11
33.0     11
69.0     10
97.0     10
45.0     10
58.0     9 
54.0     9 
68.0     9 
61.0     9 
56.0     8 
66.0     8 
71.0     8 
57.0     8 
65.0     7 
62.0     7 
38.0     5 
52.0     5 
51.0     5 
47.0     5 
64.0     5 
55.0     5 
53.0     4 
59.0     4 
41.0     4 
42.0     4 
39.0     3 
37.0     3 
29.0     3 
20.0     3 
48.0     3 
46.0     3 
44.0     3 
40.0     3 
31.0     2 
35.0     2 
17.0     2 
27.0     2 
34.0     2 
43.0     2 
0.0      2 
25.0     2 
13.0     1 
NaN      1 
99.0     1 
49.0     1 
22.0     1 
23.0     1 
Name: CoverageRatio%, dtype: int64

___________________________________ 



In [5]:
# 36 greatest people saw 0 cars  in each week, count values top 5

for week in df['Week'].unique():
    print('Week',week,':')
    print(df.sort_values(['PeopleSaw0Cars'], ascending=[False]).loc[(df['Week'] == week)][:36]['Hour'].value_counts()[:5])
    print('_______________________________________\n')

Week 1 :
9     6
7     6
8     5
19    3
18    3
Name: Hour, dtype: int64
_______________________________________

Week 2 :
8     6
18    5
10    5
9     5
7     4
Name: Hour, dtype: int64
_______________________________________

Week 3 :
23    3
19    3
7     3
10    2
3     2
Name: Hour, dtype: int64
_______________________________________

Week 4 :
18    6
19    5
9     4
8     3
14    2
Name: Hour, dtype: int64
_______________________________________

Week 5 :
8     5
9     5
18    4
19    3
20    3
Name: Hour, dtype: int64
_______________________________________



In [6]:
df.sort_values(['PeopleSaw0Cars'], ascending=[False]).loc[df['PeopleSaw0Cars']>10]['Hour'].value_counts()[:3]

8     26
18    23
9     23
Name: Hour, dtype: int64

In [8]:
print('a:')
for week in df['Week'].unique():
    dftarget = df.sort_values(['Demand'], ascending=[False]).loc[(df['Week'] == week) ][:36]
    print('Average driver\'s net income per hour in Week',week,'before tripling is',round(dftarget['RidesPerOnline'].mean(),2)
          *8,'euros')

print('---------------------------------------')
print('b(1):')  
print('If RPH is 3 without increasing passengers, then we can Decrease our drivers by the following rate which increases the remained drivers\'s income triple:')
for week in df['Week'].unique():
    dftarget = df.sort_values(['Demand'], ascending=[False]).loc[(df['Week'] == week) ][:36]
    newdriver = dftarget['Rides'].mean()/3/(dftarget['OnlinePerDrivers'].mean())
    print('Week',week,':',round((newdriver-dftarget['Drivers'].mean())/dftarget['Drivers'].mean()*100,1),'%')
print('\n''b(2):')
print('Now, let\'s multiply 3 in a results:')
for week in df['Week'].unique():
    dftarget = df.sort_values(['Demand'], ascending=[False]).loc[(df['Week'] == week) ][:36]
    print('Average driver\'s net income per hour in Week',week,'after tripling finished rides is',round(dftarget['RidesPerOnline'].mean()*8*3,2),'euros')
    
print('---------------------------------------')
print('c:')   
for week in df['Week'].unique():
    dftarget = df.sort_values(['Demand'], ascending=[False]).loc[(df['Week'] == week) ][:36]
    print('In average, We need',round(dftarget['PeopleSaw0Cars'].mean()/dftarget['PeopleSaw+1Cars'].mean()*60,0),'minutes ,ore per each driver in week',week,'to fulfil the missed demand')


a:
Average driver's net income per hour in Week 1 before tripling is 6.72 euros
Average driver's net income per hour in Week 2 before tripling is 6.08 euros
Average driver's net income per hour in Week 3 before tripling is 6.16 euros
Average driver's net income per hour in Week 4 before tripling is 9.12 euros
Average driver's net income per hour in Week 5 before tripling is 10.72 euros
---------------------------------------
b(1):
If RPH is 3 without increasing passengers, then we can Decrease our drivers by the following rate which increases the remained drivers's income triple:
Week 1 : -72.4 %
Week 2 : -76.1 %
Week 3 : -74.7 %
Week 4 : -63.2 %
Week 5 : -57.1 %

b(2):
Now, let's multiply 3 in a results:
Average driver's net income per hour in Week 1 after tripling finished rides is 20.15 euros
Average driver's net income per hour in Week 2 after tripling finished rides is 18.24 euros
Average driver's net income per hour in Week 3 after tripling finished rides is 18.44 euros
Average d

In [21]:
# 36 lowest Supply/Demand  in each week, count values top 5. Is it a good determinant od undersupply or not?

for week in df['Week'].unique():
    print('Week',week,':')
    print(df.sort_values(['SupplyDemand'], ascending=[True]).loc[(df['Week'] == week)][:36]['Hour'].value_counts()[:5])
    print('_______________________________________\n')

Week 1 :
18    5
9     5
8     5
7     5
19    4
Name: Hour, dtype: int64
_______________________________________

Week 2 :
9     5
8     5
18    4
23    3
10    3
Name: Hour, dtype: int64
_______________________________________

Week 3 :
0     4
19    3
23    3
20    2
22    2
Name: Hour, dtype: int64
_______________________________________

Week 4 :
18    6
19    5
9     5
14    2
1     2
Name: Hour, dtype: int64
_______________________________________

Week 5 :
9     6
18    4
8     4
20    3
19    3
Name: Hour, dtype: int64
_______________________________________



In [21]:
# Rush hours (8,9,18,19) considering all 5 weeks

df2 = df.loc[(df['Hour']==8)|(df['Hour']==9)|(df['Hour'] ==18)|(df['Hour']==19)]
dfresult = pd.DataFrame(data={'Drivers':[''],},columns=['Drivers'])

dfresult['Drivers'] = round(df2['Drivers'].mean(),0)
dfresult['Online(H)'] =round(df2['Online(H)'].mean(),0)
dfresult['OnlinePerDrivers'] =round(df2['OnlinePerDrivers'].mean(),1)
dfresult['Booking(H)'] =round(df2['Booking(H)'].mean(),0)
dfresult['Waiting(H)'] =round(df2['Waiting(H)'].mean(),0)
dfresult['Rides'] =round(df2['Rides'].mean(),0)
dfresult['RidesPerOnline'] =round(df2['RidesPerOnline'].mean(),2)
dfresult['PeopleSaw0Cars'] =round(df2['PeopleSaw0Cars'].mean(),0)
dfresult['PeopleSaw+1Cars'] =round(df2['PeopleSaw+1Cars'].mean(),0)
dfresult['Demand'] =round(df2['Demand'].mean(),0)
dfresult['CoverageRatio%'] =round(df2['CoverageRatio%'].mean(),0)
dfresult['Supply'] =round(dfresult['Waiting(H)']/dfresult['OnlinePerDrivers'].mean(),0)
dfresult['SupplyDemand'] =round(dfresult['Supply']/dfresult['Demand'].mean(),1)

dfresult

Drivers  Online(H)  OnlinePerDrivers  Booking(H)  Waiting(H)  Rides  \
0  71.0     28.0       0.4               12.0        16.0        22.0    

   RidesPerOnline  PeopleSaw0Cars  PeopleSaw+1Cars  Demand  CoverageRatio%  \
0  0.81            26.0            53.0             78.0    71.0             

   Supply  SupplyDemand  
0  40.0    0.5

**Approach 1 (Supply-friendly):** RPH(33/28) multiply 8 eur = 9.42 eur but passengers should wait for 29 minutes more or shift, 45% increase of income. user disatisfaction, driver satisfaction

**Approach 2 (Demand-friendly):** RPH(0.81) multiply 8 eur = 6.48 eur but drivers will lose money in non-rush hours, aound 50% increase in drivers means increase in Online(H) in hours that we have oversupply. Then RidesPerOnine will decrease, so the drivers income.